crwatcher/watcher.go

```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
 	cw.handler = cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.Filter != "" && !cw.hasFilter(r) {
+				return
+			}
 			con.ResourceAdded(r)
 		},
 		DeleteFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.Filter != "" && !cw.hasFilter(r) {
+				return
+			}
 			con.ResourceDeleted(r)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldR := oldObj.(*unstructured.Unstructured)
 			newR := newObj.(*unstructured.Unstructured)
-			con.ResourceUpdated(oldR, newR)
+			if cw.Config.Filter != "" {
+				cw.filteredUpdate(con, oldR, newR)
+			} else {
+				con.ResourceUpdated(oldR, newR)
+			}
 		},
 	}
 }
 
+// filteredUpdate performs an updated based on the filter status of the given resource. In the case that we have a
+// filter match for the old resource, but not for the new, we should delete the old resource. Either a no-op or delete
+// is valid here, but we've decided on doing a delete.
+func (cw *CRWatcher) filteredUpdate(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
```

Comment:
This method is kind of hard to follow and I know what it should be doing.  Can we make it look something like
```
if cw.clearsFilter(newR) {
    con.ResourceUpdated(oldR, newR)
  } else if cw.clearsFilter(oldR) {
    // The resource is filtered out and needs to be removed as it previously existed
    con.ResourceDeleted(oldR)
  }
```
This might not be better for readability so I'm open to suggestions.

Reply:
I agree, I actually hadn't seen this until I re-organized and came up with the same exact thing. To me that suggests it's the proper approach :)